### PR TITLE
Updated Lift Spec (1/x)

### PR DESCRIPTION
**Pull request by matthewmueller**
[Rendered View](https://github.com/prisma/specs/blob/lift/lift/Readme.md)

Rewrote the Lift spec to hopefully unblock planning. Let me know if you want me to flush out some more specific areas. 

There's still lots of TODOs, I'm not as intimately familiar with Lift as @timsuchanek and @mavilein, but hopefully this will give us a basis to discuss improvements going forward.

Part of https://github.com/prisma/specs/issues/156

### CODE REVIEW

File: lift/Readme.md [hunk 1]
```diff
@@ -6,271 +6,360 @@
...
-migrations. Therefore one can call the imperative migrations also **manual migrations**.
+### Migration
 
-### Datamodel Snapshots
+A Migration is a grouping of one or more Steps. Migrations run in a transaction if the datasource allows it.
 
-As mentioned in [Declarative migration](#how-does-naming-of-migrations-work), the user adjusts the `datamodel.prisma` file, which results in a new migration
-folder when running `prisma migrate`. In order for the migration engine to know which changes need to be applied, it needs a "copy" of the datamodel. As this
-"copy" describes the datamodel in a specific point in time, we also refer to it as a "snapshot".
+### Hook
```
Comment (janpio) [inline lift/Readme.md]: Do we actually, really already have Migration Hooks?

File: lift/Readme.md [hunk 2]
```diff
@@ -6,271 +6,360 @@
...
-history are the following:
+Lift has 2 parts: a **Lift Client** and a **Lift Server**. By default, the Lift Server runs locally as a sidecar process. This makes it easy to get started. As
+your team's data requirements become more complex, you may prefer to handle your migrations on a remote machine where you have fine-grained access control over
+migrations.
 
-- Prisma can make sure, that the local migrations are in sync with the databases migrations
-- Prisma can ensure, that migrations don't get run twice
-- Prisma can give an overview in a GUI interface about the currently running migration and migrations that already have been running
-- Prisma can rollback migrations when there has been an error
+```
```
Comment (janpio) [inline lift/Readme.md]: Cosmetics: This diagram is a bit messed up on GH. Is this fixable?

File: lift/Readme.md [hunk 3]
```diff
@@ -6,271 +6,360 @@
...
+- Add specific database primitives like functions in Postgres, which are not yet supported by Prisma.
+- Seed the database with data after a migration has been executed.
 
-### Migration History
+## Architecture
 
-When a migration is running or has been executed successfully, Prisma writes this information into a table or collection in the database. This table is also
-known as the "Migration History". It includes all successful and failed migrations of the specific Prisma Project. The benefits of storing this migration
-history are the following:
+Lift has 2 parts: a **Lift Client** and a **Lift Server**. By default, the Lift Server runs locally as a sidecar process. This makes it easy to get started. As
```
Comment (janpio) [inline lift/Readme.md]: Valid to link "sidecar process" to e.g. https://blog.davemdavis.net/2018/03/13/the-sidecar-pattern/ for context / noobs like me?

File: lift/Readme.md [hunk 4]
```diff
@@ -6,271 +6,360 @@
...
+your team's data requirements become more complex, you may prefer to handle your migrations on a remote machine where you have fine-grained access control over
```
Comment (janpio) [inline lift/Readme.md]: Do we already support remote migration servers?

File: lift/Readme.md [hunk 5]
```diff
@@ -6,271 +6,360 @@
...
+    └─ schema.prisma
+    └─ README.md
 ```
 
-Let's discuss a few questions.
+A migration contains 3 files:
 
-### Naming of migration folders
+- **steps.json:** contains a JSON list of steps to run against the database
+- **schema.prisma:** contains a snapshot of your `schema.prisma` file at a specific point in time.
```
Comment (janpio) [inline lift/Readme.md]: before or after the migration?

File: lift/Readme.md [hunk 6]
```diff
@@ -6,271 +6,360 @@
...
+
+Up runs the unapplied `up` scripts in ascending order by the migration's timestamp. For example, if we have the following migrations directory:
+
+```
+migrations/
+  └─ 20190920142118-initial/
+    └─ steps.json
+    └─ schema.prisma
+    └─ README.md
+    └─ before.sh
```
Comment (janpio) [inline lift/Readme.md]: `before.sh` and `after.sh` are a new concept here (or did I miss it?). I assume those are hooks? If so, and these are the common names for them, this should be explained in the "Hooks" sections.

File: lift/Readme.md [hunk 7]
```diff
@@ -6,271 +6,360 @@
...
+    └─ steps.json
+    └─ schema.prisma
+    └─ README.md
 ```
 
-Let's discuss a few questions.
+A migration contains 3 files:
 
-### Naming of migration folders
+- **steps.json:** contains a JSON list of steps to run against the database
```
Comment (janpio) [inline lift/Readme.md]: Both the up and down versions? Or are those the sames, and the different result is just calculated when executing up or down?

File: lift/Readme.md [hunk 8]
```diff
@@ -6,271 +6,360 @@
...
+5. `20190920142118-initial/steps.json`
+6. `20190920142118-initial/before.sh`
 
-There are a few use-cases when you want to run your own custom migration logic:
+### Lift Server
 
-- Making sure that the data adheres to constraints that you want to introduce, like `unique` or `non null` constraints
-- Adding specific database primitives like functions in Postgres, which are not yet supported by Prisma
-- Initializing data after a migration has been executed. Both SQL or the Prisma Client could be useful here
+The Lift server is a low-level interface that the Lift Client communicates with. Currently the client communicates to the migration server in the JSONRPC format
```
Comment (janpio) [inline lift/Readme.md]: migration server = lift server?

File: lift/Readme.md [hunk 9]
```diff
@@ -6,271 +6,360 @@
...
 
-As mentioned in the [basic example](#basic-example), a simple `prisma migrate` is needed to migrate the database and create a new migration.
+The Lift Client is built into the Prisma 2 CLI but is also accessible programmatically in the Prisma SDK. The Lift Client has 3 methods:
 
-These are the commands:
+#### Save
 
-```bash
-prisma migrate                                Creates a migration, clears the draft and migrates the database
+Save method takes the difference of your schema file with the current state of your datasources and creates a migration folder. This migration folder contains
```
Comment (janpio) [inline lift/Readme.md]: migration vs. migration folder. Is it the same thing? Maybe define.

File: lift/Readme.md (hunk 1 above)
Comment (mavilein) [reply lift/Readme.md]: No

File: lift/Readme.md (hunk 4 above)
Comment (mavilein) [reply lift/Readme.md]: No

File: lift/Readme.md (hunk 5 above)
Comment (mavilein) [reply lift/Readme.md]: It is the schema you used when calling `prisma lift save`. So it is the state after the migration.

File: lift/Readme.md (hunk 7 above)
Comment (mavilein) [reply lift/Readme.md]: They are the same. We calculate up and down on the fly.

File: lift/Readme.md [hunk 10]
```diff
@@ -6,271 +6,360 @@
...
 
-With imperative migrations, the user needs to both calculate the "diff" in his head and write the needed actions down, which will result in the new schema.
+</details>
 
-While in Prismas model the declarative datamodel is the source of truth for migrations, Active Record takes the imperative migrations as the source of truth for
-migrations. Therefore one can call the imperative migrations also **manual migrations**.
+### Migration
 
-### Datamodel Snapshots
+A Migration is a grouping of one or more Steps. Migrations run in a transaction if the datasource allows it.
```
Comment (mavilein) [inline lift/Readme.md]: This should be marked as not implemented.

File: lift/Readme.md [hunk 11]
```diff
@@ -6,271 +6,360 @@
...
 
-As mentioned in [Declarative migration](#how-does-naming-of-migrations-work), the user adjusts the `datamodel.prisma` file, which results in a new migration
-folder when running `prisma migrate`. In order for the migration engine to know which changes need to be applied, it needs a "copy" of the datamodel. As this
-"copy" describes the datamodel in a specific point in time, we also refer to it as a "snapshot".
+### Hook
 
-### Migration Engine
+A hook is a custom shell script that runs either before or after the migration. Hooks give you full control over your migrations. You can write hooks to:
```
Comment (mavilein) [inline lift/Readme.md]: It does not really give you full control imo. I would phrase it that it allows you to customise a migration when the normal migrations don't cut it.

File: lift/Readme.md (hunk 8 above)
Comment (mavilein) [reply lift/Readme.md]: Should we actually call those servers? Or should we rather embrace our internal engine term?

File: lift/Readme.md [hunk 12]
```diff
@@ -6,271 +6,360 @@
...
-| ---------- | ------------- |
-| after.sh   | sh after.sh   |
-| after.bash | bash after.sh |
-| after      | ./after       |
-| after.js   | ./after.js    |
-| after.ts   | ./after.ts    |
+#### unapplyMigration
 
-### Up & Down in migration scripts
+Unapplies the Steps in the migrations folder. This is called by the Down method in the Lift Client.
```
Comment (mavilein) [inline lift/Readme.md]: This is not entirely correct. `unapplyMigration` just rolls back the last migration in the database. This method is called repeatedly by the CLI.

File: lift/Readme.md [hunk 13]
```diff
@@ -6,271 +6,360 @@
...
-- Having SQL / other database native escape hatches to perform structural database changes which Prisma doesn't support yet
+- A project has many migrations
+- A migration has many steps
+- A migration has many hooks
 
-## Terminology
+### Step
 
-Assuming that the reader knows, what Prisma 1.x looks like, the following terms are new concepts which we introduce with this spec.
+A Step is a single command that we'll run against the database. Steps describe models, fields and relationships generically, so they can be easily translated to
```
Comment (mavilein) [inline lift/Readme.md]: This sounds like as if one step would always lead to one command against the database. But this is not the case. Sometimes a step does not even trigger a single database command, e.g. `CreateEnum` which does not have a representation in the database at the moment. Often multiple steps are grouped into one command, e.g. a `CreateModel` and multiple `CreateField` result in one `CREATE TABLE` in the sql connector.

**Comment by janpio** [comment]
A list of things that were present in the past, but are not any more:

1. Imperative vs. Declarative Migration: Pretty much fluff, not really necessary.
2. Migration Engine: Technical implementation detail?
3. Migration History: Seems like this could be important.
4. Migration scripts: Now called hooks? Missing a few details.

**Comment by janpio** [comment]
Thought: Current `prisma2 dev` might also be considered part of Lift and might be relevant here.

**Comment by matthewmueller** [comment]
Thanks for reviewing! I've incorporated the feedback. 

@sorenbs, good point I've also replaced that section with 

> ### How can you rename a model in Lift?
> 
> To rename a field, you'll add an `UpdateModel` step in your **steps.json** file directly.

File: lift/Readme.md [hunk 14]
```diff
@@ -245,9 +253,11 @@ migrations/
 
 A migration folder contains 3 files:
 
-- **steps.json:** contains a JSON list of steps to run against the database
-- **schema.prisma:** contains a snapshot of your `schema.prisma` file at a specific point in time.
+- **steps.json:** contains a JSON list of steps to run against the database. Steps contains only the up steps, the down steps are calculated on the fly.
```
Comment (janpio) [inline lift/Readme.md]: Is that technically correct @mavilein? Or is this more general and both steps are generated?

File: lift/Readme.md [hunk 15]
```diff
@@ -245,9 +253,11 @@ migrations/
...
+- **schema.prisma:** contains a snapshot of your `schema.prisma` file at a specific point in time after the migration has occurred.
 - **README.md:** contains information about the migration. Includes the underlying raw commands (e.g. SQL) that run against the datasource.
+- **before.sh:** optional hook you can run before migrating your schema.
```
Comment (janpio) [inline lift/Readme.md]: Will this be limited to `.sh` files or will it also enable `.js` and similar?

File: lift/Readme.md (hunk 1 above)
Comment (janpio) [reply lift/Readme.md]: This is now covered: https://github.com/prisma/specs/blob/lift/lift/Readme.md#hook

File: lift/Readme.md (hunk 4 above)
Comment (janpio) [reply lift/Readme.md]: Currently still present in first paragraph here: https://github.com/prisma/specs/blob/lift/lift/Readme.md#architecture If we want to turn the "remote" stuff into a future thing, then it maybe should be removed here as well.

File: lift/Readme.md (hunk 5 above)
Comment (janpio) [reply lift/Readme.md]: Fixed: https://github.com/prisma/specs/blob/lift/lift/Readme.md#save

File: lift/Readme.md [hunk 16]
```diff
@@ -2,275 +2,378 @@
...
 ```
 
-Let's discuss a few questions.
+A migration folder contains 3 files:
 
-### Naming of migration folders
+- **steps.json:** contains a JSON list of steps to run against the database. Steps contains only the up steps, the down steps are calculated on the fly.
+- **schema.prisma:** contains a snapshot of your `schema.prisma` file at a specific point in time after the migration has occurred.
+- **README.md:** contains information about the migration. Includes the underlying raw commands (e.g. SQL) that run against the datasource.
+- **before.sh:** optional hook you can run before migrating your schema.
```
Comment (janpio) [inline lift/Readme.md]:
```suggestion
- **before.sh:** optional [hook](#hook) you can run before migrating your schema.
```

File: lift/Readme.md [hunk 17]
```diff
@@ -2,275 +2,378 @@
...
+- **after.sh:** optional hook you can run after migrating your schema.
```
Comment (janpio) [inline lift/Readme.md]:
```suggestion
- **after.sh:** optional [hook](#hook) you can run after migrating your schema.
```

File: lift/Readme.md [hunk 18]
```diff
@@ -2,275 +2,378 @@
...
 
-When executing `prisma migrate` without providing any args, a migration with the name `yyyyMMddHHmmss` (e.g. `20190322092247`) will be created. Especially when
-working in a team, it can be very useful to add a custom name to the migration. A name can be provided like this:
+**Note:** `migrate save` does not run migrations, it simply creates them.
```
Comment (janpio) [inline lift/Readme.md]: What does `migrate save` refer to here? CLI command?

File: lift/Readme.md (hunk 14 above)
Comment (mavilein) [reply lift/Readme.md]: It's a little bit more involved. The `steps.json` only represents the changes that should be applied to the current datamodel to deduct the desired datamodel. Based on the current and desired datamodel we generate the actual database steps that will be performed for up and down.
To put it into equations:
1. `desired_datamodel = current_datamodel + steps.json`
2. `actual_database_steps = generate(current_datamodel, desired_datamodel)` <- this function is connector specific

Does that explanation help?

Comment (janpio) [reply lift/Readme.md]: It does and gives me details, thanks! And it also kind of confirms that "up steps" is valid enough here in what @matthewmueller wrote - it contains things to go from current to desired, which I think is fine to describe as "up steps", right?

Comment (mavilein) [reply lift/Readme.md]: Yes i think it's accurate enough.

File: lift/Readme.md [hunk 19]
```diff
@@ -2,275 +2,378 @@
...
+- Add specific database primitives like functions in Postgres, which are not yet supported by Prisma.
+- Seed the database with data after a migration has been executed.
 
-## Detailed design
+## Architecture
 
-As mentioned in the [basic example](#basic-example), a simple `prisma migrate` is needed to migrate the database and create a new migration.
+Lift has 2 parts: a **Lift Client** and a **Lift Engine**. By default, the Lift Engine runs locally as a
+[sidecar process](https://blog.davemdavis.net/2018/03/13/the-sidecar-pattern/). This makes it easy to get started. As your team's data requirements become more
+complex, you may prefer to handle your migrations on a remote machine where you have fine-grained access control over migrations.
```
Comment (janpio) [inline lift/Readme.md]: "remote machine" / remote migration engine is not implemented yet and was removed elsewhere in the spec.

File: lift/Readme.md (hunk 4 above)
Comment (janpio) [reply lift/Readme.md]: Now commented at https://github.com/prisma/specs/pull/159/files#r328962349

**Comment by janpio** [comment]
As we talked about the actual RPC commands under "Lift Engine" should be moved to a new issue that talks about making these commands a public API that should be stable. It would probably also make sense to have these in a `.md` file in `prisma/lift` - I am sure @tomhoule can help out there.

File: lift/Readme.md (hunk 2 above)
Comment (matthewmueller) [reply lift/Readme.md]: Screen Shot 2019-09-27 at 11 06 38 AM

^ seems okay to me.

File: lift/Readme.md (hunk 6 above)
Comment (matthewmueller) [reply lift/Readme.md]: good call, it's defined now

File: lift/Readme.md (hunk 12 above)
Comment (matthewmueller) [reply lift/Readme.md]: good call

File: lift/Readme.md [hunk 20]
```diff
@@ -2,275 +2,424 @@
...
+
+Same options and arguments as `prisma2 lift up`
+
+### Other Commands
+
+Lift functionality is also part of other commands:
+
+#### `prisma2 dev`
+
+TODO
```
Comment (janpio) [inline lift/Readme.md]: Can someone suggest a wording of what the Lift parts of `prisma2 dev` actually are and how they work?

Comment (janpio) [reply lift/Readme.md]: Ping @timsuchanek @tomhoule

**Comment by janpio** [comment]
What was the motivation of replacing the generalized spec of what the CLI should do to pasting the documentation? Isn't that the wrong way around?

PS: Content of the last two commits is kind of messed up :/

**Pull request merged**